packages/astro/src/vite-plugin-markdown/images.ts

```diff
@@ -1,34 +1,43 @@
 export type MarkdownImagePath = { raw: string; resolved: string; safeName: string };
 
 export function getMarkdownCodeForImages(imagePaths: MarkdownImagePath[], html: string) {
-	return `
-		import { getImage } from "astro:assets";
-		${imagePaths
-			.map((entry) => `import Astro__${entry.safeName} from ${JSON.stringify(entry.raw)};`)
-			.join('\n')}
+  return `
+    import { getImage } from "astro:assets";
+    ${imagePaths
+      .map((entry) => `import Astro__${entry.safeName} from ${JSON.stringify(entry.raw)};`)
+      .join('\n')}
 
-		const images = async function() {
-			return {
-				${imagePaths
-					.map((entry) => `"${entry.raw}": await getImage({src: Astro__${entry.safeName}})`)
-					.join(',\n')}
-				}
-			}
+    const images = async function(html) {
+      const imageSources = {};
+      ${imagePaths
+        .map((entry) => {
+          return `{
+            const regex = new RegExp('__ASTRO_IMAGE_="([^"]*' + ${JSON.stringify(entry.raw)} + '[^"]*)"', 'g');
+            const match = regex.exec(html);
+            if (match) {
+              const imageProps = JSON.parse(match[1].replace(/&#x22;/g, '"'));
+              const { src, ...props } = imageProps;
+              imageSources[${JSON.stringify(entry.raw)}] = await getImage({src: Astro__${entry.safeName}, ...props});
+            }
+          }`;
+        })
+        .join('\n')}
+      return imageSources;
+    }
 
-			async function updateImageReferences(html) {
-				return images().then((images) => {
-					return html.replaceAll(/__ASTRO_IMAGE_="([^"]+)"/gm, (full, imagePath) =>
-					spreadAttributes({
-						src: images[imagePath].src,
-						...images[imagePath].attributes,
-					})
-					);
-				});
-			}
-
-			// NOTE: This causes a top-level await to appear in the user's code, which can break very easily due to a Rollup
-			// bug and certain adapters not supporting it correctly. See: https://github.com/rollup/rollup/issues/4708
-			// Tread carefully!
-			const html = await updateImageReferences(${JSON.stringify(html)});
-		`;
+    async function updateImageReferences(html) {
+      return images(html).then((imageSources) => {
+        return html.replaceAll(/__ASTRO_IMAGE_="([^"]+)"/gm, (full, imagePath) => {
+          const decodedImagePath = JSON.parse(imagePath.replace(/&#x22;/g, '"'));
+          return spreadAttributes({
+            src: imageSources[decodedImagePath.src].src,
+            ...imageSources[decodedImagePath.src].attributes,
+          });
+        });
+      });
+    }
+    
+    const html = await updateImageReferences(${JSON.stringify(html)});
```

Comment:
```suggestion
	  // NOTE: This causes a top-level await to appear in the user's code, which can break very easily due to a Rollup
	  // bug and certain adapters not supporting it correctly. See: https://github.com/rollup/rollup/issues/4708
	  // Tread carefully!
    const html = await updateImageReferences(${JSON.stringify(html)});
```

Comment here is quite important, this `await` being there can break the build randomly with no error messages due to the Rollup bug.

Reply:
Ah yeah sorry I didn't mean to delete